src/Xamarin.Android.Build.Tasks/Tasks/CollectNativeFilesForArchive.cs

```diff
@@ -0,0 +1,325 @@
+				LogSanitizerError ($"Unable to map Android ABI {abi} to clang ABI");
+				return;
+			}
+
+			string sanitizerLib = $"libclang_rt.{sanitizerName}-{clangAbi}-android.so";
+			string sanitizerLibPath = Path.Combine (clangDir, sanitizerLib);
+			if (!File.Exists (sanitizerLibPath)) {
+				LogSanitizerError ($"Unable to find sanitizer runtime for the {CheckedBuild} checker at {sanitizerLibPath}");
+				return;
+			}
+
+			AddNativeLibrary (files, sanitizerLibPath, abi, sanitizerLib);
+		}
+	}
+
+	string? GetNativeLibraryAbi (ITaskItem lib)
+	{
+		// If Abi is explicitly specified, simply return it.
+		var lib_abi = AndroidRidAbiHelper.GetNativeLibraryAbi (lib);
+
+		if (string.IsNullOrWhiteSpace (lib_abi)) {
+			Log.LogCodedError ("XA4301", lib.ItemSpec, 0, Properties.Resources.XA4301_ABI, lib.ItemSpec);
+			return null;
+		}
+
+		return lib_abi;
+	}
+
+	void AddNativeLibraries (ArchiveFileList files, string [] supportedAbis, System.Collections.Generic.IEnumerable<LibInfo> libs)
+	{
+		if (libs.Any (lib => lib.Abi == null))
+			Log.LogCodedWarning (
+					"XA4301",
+					Properties.Resources.XA4301_ABI_Ignoring,
+					string.Join (", ", libs.Where (lib => lib.Abi == null).Select (lib => lib.Path)));
+		libs = libs.Where (lib => lib.Abi != null);
+		libs = libs.Where (lib => supportedAbis.Contains (lib.Abi));
+		foreach (var info in libs) {
+			AddNativeLibrary (files, info.Path, info.Abi!, info.ArchiveFileName, info.Item);
+		}
+	}
+
+	void AddAdditionalNativeLibraries (ArchiveFileList files, string [] supportedAbis)
+	{
+		if (AdditionalNativeLibraryReferences == null || !AdditionalNativeLibraryReferences.Any ())
+			return;
+
+		var libs = AdditionalNativeLibraryReferences
+			.Select (l => new LibInfo (
+				path: l.ItemSpec,
+				link: null,
+				abi: AndroidRidAbiHelper.GetNativeLibraryAbi (l),
+				archiveFileName: l.GetMetadata ("ArchiveFileName"),
+				item: l
+			));
+
+		AddNativeLibraries (files, supportedAbis, libs);
+	}
+
+	void AddNativeLibrary (ArchiveFileList files, string path, string abi, string? archiveFileName, ITaskItem? taskItem = null)
+	{
+		string fileName = string.IsNullOrEmpty (archiveFileName) ? Path.GetFileName (path) : archiveFileName!;
+		var item = (filePath: path, archivePath: MakeArchiveLibPath (abi, fileName));
+		if (files.Any (x => x.archivePath == item.archivePath)) {
+			Log.LogCodedWarning ("XA4301", path, 0, Properties.Resources.XA4301, item.archivePath);
+			return;
+		}
+
+		ELFHelper.AssertValidLibraryAlignment (Log, ZipAlignmentPages, path, taskItem);
+		if (!ELFHelper.IsEmptyAOTLibrary (Log, item.filePath)) {
+			files.Add (item);
+		} else {
+			Log.LogDebugMessage ($"{item.filePath} is an empty (no executable code) AOT assembly, not including it in the archive");
+		}
+	}
+
+	// This method is used only for internal warnings which will never be shown to the end user, therefore there's
+	// no need to use coded warnings.
+	void LogSanitizerWarning (string message)
```

Comment:
This is probably the original code, but what is a "internal warning"? This shows the warning to the user, how is it internal?

Reply:
Seems like this should just use `LogDebugMessage()`?

Reply:
It is original code, however it looks like it is only used for "checked" builds which requires the internal property `$(_AndroidCheckedBuild)` to be set.  I updated the comment on it to mention this.